Automattic/jetpack
# [Related Posts]: Allow assets to be dequeued

Fixes #20039 

It should be possible to dequeue both the JavaScript and the CSS for the Related Posts, in order to load in your own files instead. This feature stopped working in #19615. 

Previously assets were enqueued immediately; in #19615 this was modified to avoid enqueueing assets until as late as possible, to prevent some errors when the scripts are loaded unnecessarily. Unfortunately this causes the bug, because any existing code using `wp_dequeue_script` for example will now run _before_ the assets are actually enqueued.

This PR moves the asset enqueueing back to the init action where it was previously run, and instead adds a few extra checks to prevent enqueueing scripts when they will not be needed (to prevent regressions from #19615).


#### Changes proposed in this Pull Request:

Technically:
* Assets are enqueued as early as possible again
* Checks added to prevent assets from being enqueued when they are not needed (if there is a Related Posts block in the post, or if the site is running an FSE theme)

Practically there should be no change, except that it should be possible to dequeue scripts now.


#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



#### Testing instructions:






We should test all the ways the Related Posts feature can be inserted. Enable the feature via `Settings -> Reading` (or via the Jetpack settings on a Jetpack site) and make sure your test site has at least ten published posts.

In each case keep the console open and check for errors. There are console errors if the script is loaded when it is not needed.

**These changes include manual changes to the diff. To test wpcom, apply the D63970-code to your sandbox rather than syncing.**

* Enable a FSE theme such as TT1 blocks. View a post that does NOT contain the Related Posts block on the frontend. You should not see the feature added automatically.
* Edit the site in the site editor and add a Related Posts block to the single post template. View a post on the frontend and verify you can see the feature where you inserted it.
* Edit the post and add a Related Posts block. View on the frontend and verify you can see the block in the content, as well as the block from the template.
* Enable a non-FSE theme such as TT1. View the post on the frontend and verify you can see the block in the content, and the feature was not added automatically to the footer.
* Remove the Related Posts block from the post and view on the frontend again. You should see the feature added automatically.
* Add the block via a shortcode (`[jetpack-related-posts]`) and verify that you see it on the frontend and that the feature is not added automatically.

To test dequeueing the script, I first verified I could see the related posts automatically being added to my Jurassic.tube site. Then I created and installed a minimal plugin for dequeuing, and verified that the Related Posts are no longer visible. 
* Verify that the Related posts div was still inserted into the page (open Inspector and find the `<div id="jp-relatedposts">`, but is empty (because the scripts that populate it were not loaded)
* Verify in the Inspector that the scripts were not loaded

Sample code for dequeueing:
```
<?php
/*
Plugin Name: My Test Plugin
*/
 
function wpdocs_dequeue_script() {
    wp_dequeue_script( 'jetpack_related-posts' );
}
add_action( 'wp_print_scripts', 'wpdocs_dequeue_script', 100 );
```
